Catch InvalidProtocolBufferException creating EncryptedSharedPreferences

Cheesebaron opened the pull request:
### Description of Change

<!-- Enter description of the fix in this section -->

Creating EncryptedSharedPreferences on some devices throw InvalidProtocolBufferException.

This PR catches this exception instead of throwing it to the consumer.

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes #13597

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

Cheesebaron committed:
Catch InvalidProtocolBufferException creating EncryptedSharedPreferences

rmarinho commented:
/azp run

Cheesebaron commented:
Not 100% sure, but perhaps we should remove all keys if this occurs.

jfversluis commented:
> Not 100% sure, but perhaps we should remove all keys if this occurs.

What is the situation here? When does this occur? Is there _any_ recovery scenario? If not, I'd say indeed wipe it.

mattleibow commented on src/Essentials/src/SecureStorage/SecureStorage.android.cs:
```diff
@@ -18,22 +19,28 @@ Task<string> PlatformGetAsync(string key)
 				{
 					// TODO: Use Logger here?
-					System.Diagnostics.Debug.WriteLine($"Unable to decrypt key, {key}, which is likely due to an app uninstall. Removing old key and returning null.");
-					Remove(key);
```
What happened to the removes?

Cheesebaron replied:
Yeah, I think I missed that one 🤔 Will add it back where appropriate. Thanks!

Cheesebaron committed:
Address PR review comments

Removing keys like before my changes when errors occur as they are likely due to a corruption

Cheesebaron commented:
> > Not 100% sure, but perhaps we should remove all keys if this occurs.
> 
> What is the situation here? When does this occur? Is there _any_ recovery scenario? If not, I'd say indeed wipe it.

It is a bug in AndroidX Security that Essentials is currently using, which breaks the Encrypted Shared Preferences when restored from a backup and results in a exception.

Situation should be better with a newer version according to the bug linked in the issue. However, there are no bindings for it 🥲

I asked @moljac on Discord if we could perhaps get a newer version bound, even though there is a policy on not binding pre-release stuff.

mattleibow commented:
/azp run

jfversluis reviewed (changes requested):
Build error 🙅 

> D:\a\_work\1\s\src\Essentials\src\SecureStorage\SecureStorage.android.cs(113,5): error CS0120: An object reference is required for the non-static field, method, or property 'SecureStorageImplementation.PlatformRemoveAll()' [D:\a\_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net7.0-android]

moljac commented:
> > > Not 100% sure, but perhaps we should remove all keys if this occurs.
> > 
> > 
> > What is the situation here? When does this occur? Is there _any_ recovery scenario? If not, I'd say indeed wipe it.
> 
> It is a bug in AndroidX Security that Essentials is currently using, which breaks the Encrypted Shared Preferences when restored from a backup and results in a exception.
> 
> Situation should be better with a newer version according to the bug linked in the issue. However, there are no bindings for it 🥲
> 
> I asked @moljac on Discord if we could perhaps get a newer version bound, even though there is a policy on not binding pre-release stuff.

... and I got green light to bind update

https://github.com/xamarin/AndroidX/pull/707

Cheesebaron committed:
Dont mark GetEncryptedSharedPreferences() static

Cheesebaron commented:
> Build error 🙅
> 
> > D:\a_work\1\s\src\Essentials\src\SecureStorage\SecureStorage.android.cs(113,5): error CS0120: An object reference is required for the non-static field, method, or property 'SecureStorageImplementation.PlatformRemoveAll()' [D:\a_work\1\s\src\Essentials\src\Essentials.csproj::TargetFramework=net7.0-android]

Oopsie. Fixed.

moljac commented:
published new previews

https://www.nuget.org/packages/Xamarin.AndroidX.Security.SecurityCrypto/1.1.0-alpha05
https://www.nuget.org/packages/Xamarin.AndroidX.Security.SecurityCrypto.Ktx/1.1.0-alpha05

samhouts commented:
/azp run

kirakosyan commented:
Hi 
Looks like it is still shown as fatal crash.
Should it be like this, or I misunderstood the fix?
target framework is net7.0-android

<img width="1271" alt="image" src="https://user-images.githubusercontent.com/4753738/228378461-d3ebaf6c-1530-46f8-8f22-8db8b50cb67c.png">

kirakosyan commented:
thanks for the fix, app is not crashing anymore

kirakosyan commented:
hi
I still have the same issue and app is crashing

<img width="1088" alt="image" src="https://user-images.githubusercontent.com/4753738/228382286-5e274642-7830-47d2-90f0-960fb5d6bd1c.png">

jonpryor commented:
See also:

  * https://learn.microsoft.com/en-us/answers/questions/1001662/suddenly-getting-securestorage-issues-in-maui

plequere-ms commented:
Would strongly recommend backporting this to .net 7.  There is no recovery path when it starts happening randomly, and in my experience, even catching exceptions is not sufficient - there are exceptions that bubble up the stack and cause crashes.

hartez commented:
/backport to net7.0
